### PR TITLE
Implement advanced search

### LINKED ISSUE #239: Add ability to search by field with logic keywords
Search currently searches all default fields and groups without the ability to narrow in on your goal. Logic keywords are also not utilized

This feature would allow named fields to be used in the search dialog to narrow down the search in large databases. It would also introduce the logical keywords AND and OR. Examples:

* user:jon AND url:google
* group:networks AND title:nas
* user:jon OR user:bob

### LINKED ISSUE #1501: Focus selected folder after search
(this might be an improvement)
## Expected Behavior
After doing a search, I'd like to see the contents of a folder that I select in my tree directly instead of having to press "Esc"/cancel search first.

## Current Behavior
I search for something, use that entry, then later I select a folder but the search result list remains in the main window.

## Possible Solution
Directly list contents of a folder on select.

## Steps to Reproduce (for bugs)

n/a

## Context
I'd like to improve my work's flow.

## Debug Info

KeePassXC - Version 2.2.4
Revision: 4723f66

Bibliotheken:
- Qt 5.9.1
- libgcrypt 1.8.1

Betriebssystem: macOS High Sierra (10.13)
CPU-Architektur: x86_64
Kernel: darwin 17.4.0

Aktivierte Erweiterungen:
- KeePassHTTP
- Auto-Type
- YubiKey

### PR DESCRIPTION
**Pull request by droidmonkey**
## Description

Fully implements advanced searching and other search related features/issues. Eliminates the searching  of group names and notes. End search after selecting a folder if global search is active (Fixes #1501). Also took the opportunity to clean up the ugly/messy entry view and model code surrounding searching.

Fixes #239 using the following syntax:

```[modifiers][field:]["]term["]```

Where __modifiers__ can be any combo of:
* ```-``` exclude this term from results
* ```+``` match this term exactly
* ```*``` term is handled as a regex

Where __field__ can be one of:
* title
* user[name]
* pass[word]
* url
* notes
* attr[ibute]
* attach[ment]

If field is omitted or non-existent the default is to search title, username, url, and notes.

Where __term__ can contain wildcard terms (if NOT regex) of:
* ```*``` match anything (or nothing)
* ```?``` match any __one__ character
* ```|``` logical OR

Sample search queries:
* ```user:jon url:google``` Search username for jon and url for google
* ```user:jon|hifi``` Search username for jon OR hifi
* ```+user:jon -url:google *notes:"secret note \d"``` Search username exactly jon, url cannot contain google, and notes contains secret note [digit]

## Motivation and context

Long requested feature

## How has this been tested?

Manually and with new search term parser tests.

## Types of changes

- ✅ New feature (non-breaking change which adds functionality)

## Checklist:

- ✅ I have read the **CONTRIBUTING** document. **[REQUIRED]**
- ✅ My code follows the code style of this project. **[REQUIRED]**
- ✅ All new and existing tests passed. **[REQUIRED]**
- ✅ I have compiled and verified my code with `-DWITH_ASAN=ON`. **[REQUIRED]**
- ✅ My change requires a change to the documentation and I have updated it accordingly.
- ✅ I have added tests to cover my changes.

### CODE REVIEW

File: src/core/EntrySearcher.cpp [hunk 1]
```diff
@@ -20,75 +20,163 @@
...
+        default:
+            found = term->regex.match(entry->resolvePlaceholder(entry->title())).hasMatch() ||
+                    term->regex.match(entry->resolvePlaceholder(entry->username())).hasMatch() ||
+                    term->regex.match(entry->resolvePlaceholder(entry->url())).hasMatch() ||
+                    term->regex.match(entry->notes()).hasMatch();
+        }
+
+        // Short circuit if we failed to match or we matched and are excluding this term
+        if (!found || term->exclude) {
             return false;
```
Comment (TheZ3ro) [inline src/core/EntrySearcher.cpp]: Shouldn't this be a `continue` instead? It is into a for and returning here will skip all next term in the for if the current is excluded

Comment (droidmonkey) [reply src/core/EntrySearcher.cpp]: It is a short circuit because this function only returns true if all the terms match (implicit AND of terms).

Comment (TheZ3ro) [reply src/core/EntrySearcher.cpp]: ok make sense

**Comment by yan12125** [comment]
> ```*``` term is handled as a regex

I like this idea ;-) Looking forward to reusing this for #1769

**Comment by droidmonkey** [comment]
Right forgot about that, lemme extract the wildcard scrub into the tools class

**Comment by insideClaw** [comment]
Sounds like exactly what was a deal-maker for me to start using KeePassXC, that it doesn't currently match a lot of sites. +1

File: src/core/Tools.cpp [hunk 2]
```diff
@@ -346,4 +347,32 @@ namespace Tools
...
+auto regexEscape = QRegularExpression(R"re(([-[\]{}()+.,\\\/^$#]))re");
+
+QRegularExpression convertToRegex(const QString& string, bool useWildcards, bool exactMatch, bool caseSensitive)
+{
+    QString pattern = string;
+
+    // Wildcard support (*, ?, |)
+    if (useWildcards) {
+        pattern.replace(regexEscape, "\\\\1");
+        pattern.replace("**", "*");
```
[the remark this answers is not in the document]
Comment (droidmonkey) [reply src/core/Tools.cpp]: Good point, this seems to have been a leftover from testing

**Comment by droidmonkey** [comment]
[PULL REQUEST](https://trello.com/c/ALJrj0WE/14-pull-request)

**Comment by droidmonkey** [comment]
Going to make a couple additions and then it will be ready

**Comment by intika** [comment]
@droidmonkey Great work this is very useful tested it and it seem to work as expected... 
Just one thing, the search syntax is not intuitive you should add a hint in the search box... instead of "Search..." something like "Search... (+pass:keyword)" or "Search... (+user:keyword)"

**Comment by droidmonkey** [comment]
I made a pretty awesome help pop-up. Any feedback?

search_help

**Comment by intika** [comment]
Amazing indeed... that will be useful to remember the syntax

**Comment by phoerious** [comment]
Despicable. 😉
No, great pop-up, but I would prefer if it had the QLineEdit background. And I would replace the part in brackets after the description with only "(logical AND)". Making a verb of "and" sounds rather clunky to me.

**Comment by phoerious** [comment]
Oh, and could you perhaps make a generic help pop-up widget of it? We could use that in other parts as well.

File: src/gui/widgets/PopupHelpWidget.cpp [hunk 3]
```diff
@@ -0,0 +1,99 @@
+/*
+ *  Copyright (C) 2017 KeePassXC Team <team@keepassxc.org>
```
Comment (phoerious) [inline src/gui/widgets/PopupHelpWidget.cpp]: 2018

File: src/gui/widgets/PopupHelpWidget.h [hunk 4]
```diff
@@ -0,0 +1,47 @@
+/*
+ *  Copyright (C) 2017 KeePassXC Team <team@keepassxc.org>
```
Comment (phoerious) [inline src/gui/widgets/PopupHelpWidget.h]: DItto,

File: src/gui/widgets/PopupHelpWidget.h [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+    void setPosition(Qt::Corner corner);
+
+protected:
+    bool eventFilter(QObject* obj, QEvent* event) override;
+    void showEvent(QShowEvent* event) override;
+
+private:
+    void alignWithParent();
+    QWidget* m_parentWindow;
+    QWidget* m_appWindow;
```
Comment (phoerious) [inline src/gui/widgets/PopupHelpWidget.h]: QPointers

File: src/gui/widgets/PopupHelpWidget.cpp (hunk 3 above)
Comment (droidmonkey) [reply src/gui/widgets/PopupHelpWidget.cpp]: CopyPasta!

**Comment by droidmonkey** [comment]
Done, and fixed the wording on the help popup, made background white, and fixed the border. Also squashed commits.

**Pull request merged**